util/internal/test/common/src/main/java/org/hibernate/search/util/impl/test/extension/parameterized/ParameterizedClassExtension.java

```diff
@@ -0,0 +1,188 @@
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+import org.junit.jupiter.api.extension.ReflectiveInvocationContext;
+import org.junit.jupiter.api.extension.TestTemplateInvocationContext;
+import org.junit.jupiter.api.extension.TestTemplateInvocationContextProvider;
+
+final class ParameterizedClassExtension
+		implements TestTemplateInvocationContextProvider, AfterTestExecutionCallback, ParameterResolver {
+	private List<Object[]> envArguments;
+	private int envIndex = 0;
+	private boolean envInitialized = false;
+
+	@Override
+	public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext)
+			throws ParameterResolutionException {
+		return true;
+	}
+
+	@Override
+	public Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext)
+			throws ParameterResolutionException {
+		return envArguments.get( envIndex )[parameterContext.getIndex()];
+	}
+
+	@Override
+	public boolean supportsTestTemplate(ExtensionContext context) {
+		return isAnnotated( context.getTestMethod(), ParameterizedSetup.class );
+	}
+
+	@Override
+	public Stream<TestTemplateInvocationContext> provideTestTemplateInvocationContexts(ExtensionContext context) {
+		Class<?> testClass = context.getTestClass().orElseThrow();
+		// find actual "tests" that we'll invoke via reflection:
+		List<ParameterizedTestMethodInvoker> testMethods = discover( testClass, context );
+
+		if ( testMethods.isEmpty() ) {
+			throw new IllegalStateException( "No tests to execute were found." );
+		}
+
+		envArguments = new ArrayList<>();
+		ParameterizedClassUtils.findParameters( envArguments, context, context.getRequiredTestMethod() );
+
+		return stream( spliteratorUnknownSize(
+				new Iterator<>() {
+
+					Iterator<ParameterizedTestMethodInvoker> test = testMethods.iterator();
+
+					@Override
+					public boolean hasNext() {
+						if ( test.hasNext() ) {
+							return true;
+						}
+						else {
+							envIndex++;
+							envInitialized = false;
+						}
+
+						if ( envIndex < envArguments.size() ) {
+							test = testMethods.iterator();
+							return test.hasNext();
+						}
+						return false;
+					}
+
+					@Override
+					public TestTemplateInvocationContext next() {
+						if ( envInitialized ) {
+							ParameterizedTestMethodInvoker testMethod = test.next();
+							write( context, StoreKey.TEST_TO_RUN, testMethod );
+
+							return new ParameterizedClassTestTemplateInvocationContext(
+									testMethod.getName(),
+									MethodInterceptor.TEST_EXECUTION_INTERCEPTOR
+							);
+						}
+						else {
+							return new ParameterizedClassTestTemplateInvocationContext(
+									"Setup Parameters: " + Arrays.toString( envArguments.get( envIndex ) ),
+									MethodInterceptor.SETUP_EXECUTION_INTERCEPTOR
+							);
```

Comment:
Could this be turned into a "BeforeEachCallback" extension added (manually) to every test instead?

That way:

1. The setup won't appear as a test in the IDE or surefire reports
2. A failure to run the setup will automatically mark the corresponding tests as failed (currently they get executed anyway... potentially using the previous setup).

Reply:
I thought that we might want to see the setup as a line, but if not... 😈 I'll control the setup with the method interceptor then and we'll just have a single  `ParameterizedClassTestTemplateInvocationContext` without the if/else part